README.md
# Painting Stoner

![Painting Stoner in action](https://storage.googleapis.com/olaviinha/github/ps-4fps.gif)
(Note that the fading animation is much smoother in reality than in this 4 frames per second gif.)

---

Painting Stoner is a simple browser-based tool for optimizing projector guided painting. I.e. when you use a real-life projector to project a digital image on your real-life canvas for real-life painting, whether you use projecting for making a rough sketch, learning perspectives or even attempting to paint the projected image.

I developed this tool for my own purposes after purchasing a $50 projector for sketching initial perspectives more accurately. All features are built around that cheap-ass projector's shortcomings as well as otherwise making many aspects of the process significantly more convenient.

Painting Stoner's power lies in three key features:
1. **Free perspective transformation of image.** This enables you to keystone the image to your canvas despite the projector position or what kind of keystoning features it has. Optional guidelines are present for more accurate positioning.
2. **Extended opacity control with animated auto-fading.** In addition to basic opacity slider, the Stoner includes an opacity range slider where you can select two opacity values. You can easily switch between these values when necessary by clicking anywhere on page or hitting spacebar on your keyboard. You may also enable animated auto-fading between these opacity values according to duration settings (separately specified image fade time, image on-screen time and image off-screen time). This way you will see both the end result without projection and the overlaying projection on regular (fast or slow) intervals while you paint. Additionally you may adjust cursor size and hide any small portion of the projection by cursor.
3. **Multiple relevant image adjustment controls** to get the projected image closer to what best suits your needs at any moment. Controls include projection backlight, grayscale, invert colours, threshold, contrast and brightness. With optional [backend setup](#setup), you get a few extra features and controls based on server-side image processing: details, posterize, vectorize, outlines, and indicative colour palette generated from image.

Tested only in [Google Chrome](https://chrome.google.com).

![Painting Stoner in action](https://storage.googleapis.com/olaviinha/github/painting-stoner/backendadjs2.gif)
Painting Stoner in action.

## Online demo

Demo has no backend support, but is otherwise fully functional.

https://inha.asia/dmo/painting-stoner


## How to use

1. Set up your projector to project on your canvas. Remember that this tool has excessive keystoning, so you may place your projector in a position where you normally might not – The Stoner will make up for any possible shortcomings in your projector's keystoning features.
2. Open [Painting Stoner](https://inha.asia/dmo/painting-stoner).
3. Drag any image file from your computer and drop it in the yellow square (yellow square will appear on page when you drag your file on the page). Alternatively you may directly paste an image or an image URL from your clipboard.
4. Click on the <i>Full screen</i> button to enter full screen mode.
5. Drag the big red corner dots to get the image on your canvas as you wish. Easiest way to go is to crop your image to the same aspect ratio as your canvas (using some other tools) – in this case you will simply drag the big red dots onto the corners of you canvas and you're done.
6. Adjust image and opacity settings to whatever you find most convenient for the situation at hand.
7. Start painting.

## Setup

### ...without backend support:

Place all the files of this repository on a server and you're done.

### ...with backend support:

Place all the files of this repository on a server and install the prerequisites. Backend support will provide a few additional controls based on server-side image processing. 

#### Prerequisites
- PHP
- Gimp-G'MIC
- G'MIC

#### Example installation

Installation procedure depends on your server setup, but for Apache2 running on Ubuntu, it should go _something like this_:

```
# Install PHP
sudo apt install php libapache2-mod-php
# Install Gimp-G'MIC & G'MIC: 
sudo apt install gimp-gmic gmic
# Restart Apache
sudo service apache2 restart
# Grant PHP write permission to tmp dir (used to store processed images temporarily)
sudo chown -R www-data:www-data /var/www/WHATEVER-PATH-TO/painting-stoner/tmp
```

Lastly, open `stoner.js` in a text editor and change `var gmicSupport` to `true`:
```
var gmicSupport = true;
```


CHANGELOG.md
# Change log

## 2021-07-12 v0.1.5

- Generate indicative palette when backend supported. When visible, palette is displayed in page background as well as inside projected image.
- New checkbox control: Indicative palette (visible only with backend support).
- New checkbox control: Boost palette (visible only with backend support).
- Minor fix: help box position (Windows browsers displayed scrollbars when hidden).
- Minor fix: Load timer (on some pageloads content was not displayed until refresh).

## 2021-02-14 v0.1.4

- Restore last generated images upon pageload when backend supported.
- New radio select control: Filter blend mode (normal, darken, lighten)
- New slider control: Bg image opacity, usable when 'darken' or 'lighten' blend mode selected.
- New slider control: Filter opacity.
- Better drag corners.

## 2021-01-27 v0.1.3

- New slider control: Outlines (visible only with backend support).

## 2021-01-02 v0.1.2

- Support for server-side Gimp-G'MIC image processing.
- New checkbox control: Details (visible only with backend support).
- New slider control: Posterize (visible only with backend support).
- New slider control: Vectorize (visible only with backend support).
- New slider control: Cursor size.
- Mouse cursor over the image creates a round hole in the projection.
- Updated grid.

## 2020-12-30 v0.1.1

- Now accepts image data paste.
- Now accepts image URL paste without the input field.
- Image is now draggable (in addition to freely transformable).
- New checkbox control: Invert.
- Default background now black (no backlight from projector).
- New slider control: Backlight.
- New slider control: Threshold.
- New slider control: Brightness.
- Added numeric value display to each single value slider.
- Added favicon.
- Added instructions.
- Added show/hide toggler for the controls bar.
- Added some further instructions.
- Help box is now larger and easier to read.
- New font.
- Graphic UI improvements.


palette.php
<?php

namespace Chirp;

// Original PHP code by Chirp Internet: www.chirpinternet.eu
// Please acknowledge use of this code by including this header.

class ImageSampler
{

  private $img;
  private $callback = NULL;
  private $initialized = FALSE;

  protected $percent = 5;
  protected $steps = 10;

  public $w, $h;
  public $sample_w = 0;
  public $sample_h = 0;

  public function __construct($imagefile)
  {
    if(!$this->img = imagecreatefromjpeg($imagefile)) {
      die("Error loading image: {$imagefile}");
    }
    $this->w = imagesx($this->img);
    $this->h = imagesy($this->img);
  }

  public function set_percent($percent)
  {
    $percent = intval($percent);
    if(($percent < 1) || ($percent > 50)) {
      die("Your \$percent value needs to be between 1 and 50.");
    }
    $this->percent = $percent;
  }

  public function set_steps($steps)
  {
    $steps = intval($steps);
    if(($steps < 1) || ($steps > 50)) {
      die("Your \$steps value needs to be between 1 and 50.");
    }
    $this->steps = $steps;
  }

  private function set_callback($callback)
  {
    try {
      $fn = new \ReflectionFunction($callback);
      if($fn->getNumberOfParameters() != 5) {
        throw new \ReflectionException("Invalid parameter count in callback function.  Usage: fn(int, int, int, bool) { ... }");
      }
      $this->callback = $callback;
    } catch(\ReflectionException $e) {
      die($e->getMessage());
    }
  }

  public function init()
  {
    $this->sample_w = $this->w / $this->steps;
    $this->sample_h = $this->h / $this->steps;
    $this->initialized = TRUE;
  }

  private function get_pixel_color($x, $y)
  {
    $rgb = imagecolorat($this->img, $x, $y);
    $r = ($rgb >> 16) & 0xFF;
    $g = ($rgb >> 8) & 0xFF;
    $b = $rgb & 0xFF;
    return [$r, $g, $b];
  }

  public function sample($callback = NULL)
  {
    if(!$this->initialized) {
      $this->init();
    }
    if(($this->sample_w < 2) || ($this->sample_h < 2)) {
      die("Your sampling size is too small for this image - reduce the \$steps value.");
    }

    if($callback) {
      $this->set_callback($callback);
    }

    $sample_size = round($this->sample_w * $this->sample_h * $this->percent / 100);

    $retval = [];
    for($i=0, $y=0; $i < $this->steps; $i++, $y += $this->sample_h) {
      $flag = FALSE;
      $row_retval = [];
      for($j=0, $x=0; $j < $this->steps; $j++, $x += $this->sample_w) {
        $total_r = $total_g = $total_b = 0;
        for($k=0; $k < $sample_size; $k++) {
          $pixel_x = $x + rand(0, $this->sample_w-1);
          $pixel_y = $y + rand(0, $this->sample_h-1);
          list($r, $g, $b) = $this->get_pixel_color($pixel_x, $pixel_y);
          $total_r += $r;
          $total_g += $g;
          $total_b += $b;
        }
        $avg_r = round($total_r/$sample_size);
        $avg_g = round($total_g/$sample_size);
        $avg_b = round($total_b/$sample_size);
        if($this->callback) {
          call_user_func_array($this->callback, [$avg_r, $avg_g, $avg_b, $w, !$flag]);
        }
        $row_retval[] = [$avg_r, $avg_g, $avg_b];
        $w = $pixel_x;
        $flag = TRUE;
      }
      $retval[] = $row_retval;
    }

    return $retval;
  }

}

$files = glob('tmp/*');
$img = $files[0];
$sampler = new \Chirp\ImageSampler($img);
$sampler->set_percent(15);
$sampler->set_steps(7);
$sampler->init();
$colors = array();

$sampler_callback = function($r, $g, $b, $w, $new_row) {
  global $colors;
  $color = $r . "," . $g . "," . $b;
  $entry = array(
    'w' => $w,
    'color' => $color
  );
  array_push($colors, $entry);
};

$sampler->sample($sampler_callback);

function getBright($palette, $brightness_limit=400) {

  global $max;
  global $diff_limit;

  $bright_colors = array();
  $filtered_bright_colors = array();
  $existing_sums = array();
  $ci = 0;

  foreach($palette as $i => $color) {
    $xcolor = explode(",", $color['color']);
    $sum = array_sum($xcolor);
    if(!in_array($sum, $existing_sums)) array_push($existing_sums, $sum);
  }

  $average = array_sum($existing_sums)/count($existing_sums);
  $brightness_min = $average * .1;
  $brightness_max = $average * .9;

  foreach($palette as $i => $color) {
    $xcolor = explode(",", $color['color']);
    $sum = array_sum($xcolor);

    if(!in_array($sum, $existing_sums)) array_push($existing_sums, $sum);
    if($sum >= $brightness_min && $sum <= $brightness_max) {

      foreach($existing_sums as $existing_sum){
        if( $existing_sum > $sum+$comp && $exiting_sum < $sum-$comp && !in_array($color, $bright_colors)){
          $color_entry = array(
            'w' => $color['w'],
            'color' => $color['color']
          );
          array_push($bright_colors, $color_entry);
          $ci++;
        }
      }
    }

  }

  $final_list = $bright_colors;
  sort($final_list);
  return $final_list;
}

echo json_encode( getBright($colors, 60) );

?>


persform.js
(function(a){a(function(){function b(e,f,c,d){this.element=e;this.style=e.style;this.computedStyle=window.getComputedStyle(e);this.width=f;this.height=c;this.useBackFacing=!!d;this.topLeft={x:0,y:0};this.topRight={x:f,y:0};this.bottomLeft={x:0,y:c};this.bottomRight={x:f,y:c}}b.useDPRFix=false;b.dpr=1;b.prototype=(function(){var d={stylePrefix:""};var f;var k;var c;var m=[[0,0,1,0,0,0,0,0],[0,0,1,0,0,0,0,0],[0,0,1,0,0,0,0,0],[0,0,1,0,0,0,0,0],[0,0,0,0,0,1,0,0],[0,0,0,0,0,1,0,0],[0,0,0,0,0,1,0,0],[0,0,0,0,0,1,0,0]];var i=[0,0,0,0,0,0,0,0];function n(){var o=document.createElement("div").style;d.stylePrefix="webkitTransform" in o?"webkit":"MozTransform" in o?"Moz":"msTransform" in o?"ms":"";f=d.stylePrefix+(d.stylePrefix.length>0?"Transform":"transform");c="-"+d.stylePrefix.toLowerCase()+"-transform-origin"}function h(){var p=this.topLeft.x-this.topRight.x;var o=this.topLeft.y-this.topRight.y;if(Math.sqrt(p*p+o*o)<=1){return true}p=this.bottomLeft.x-this.bottomRight.x;o=this.bottomLeft.y-this.bottomRight.y;if(Math.sqrt(p*p+o*o)<=1){return true}p=this.topLeft.x-this.bottomLeft.x;o=this.topLeft.y-this.bottomLeft.y;if(Math.sqrt(p*p+o*o)<=1){return true}p=this.topRight.x-this.bottomRight.x;o=this.topRight.y-this.bottomRight.y;if(Math.sqrt(p*p+o*o)<=1){return true}p=this.topLeft.x-this.bottomRight.x;o=this.topLeft.y-this.bottomRight.y;if(Math.sqrt(p*p+o*o)<=1){return true}p=this.topRight.x-this.bottomLeft.x;o=this.topRight.y-this.bottomLeft.y;if(Math.sqrt(p*p+o*o)<=1){return true}return false}function j(q,p,o){return q.x*p.y+p.x*o.y+o.x*q.y-q.y*p.x-p.y*o.x-o.y*q.x}function l(){var p=j(this.topLeft,this.topRight,this.bottomRight);var o=j(this.bottomRight,this.bottomLeft,this.topLeft);if(this.useBackFacing){if(p*o<=0){return true}}else{if(p<=0||o<=0){return true}}var p=j(this.topRight,this.bottomRight,this.bottomLeft);var o=j(this.bottomLeft,this.topLeft,this.topRight);if(this.useBackFacing){if(p*o<=0){return true}}else{if(p<=0||o<=0){return true}}return false}function g(){if(h.apply(this)){return 1}if(l.apply(this)){return 2}return 0}function e(){var x=this.width;var v=this.height;var E=0;var D=0;var t=this.computedStyle.getPropertyValue(c);if(t.indexOf("px")>-1){t=t.split("px");E=-parseFloat(t[0]);D=-parseFloat(t[1])}else{if(t.indexOf("%")>-1){t=t.split("%");E=-parseFloat(t[0])*x/100;D=-parseFloat(t[1])*v/100}}var G=[this.topLeft,this.topRight,this.bottomLeft,this.bottomRight];var q=[0,1,2,3,4,5,6,7];for(var B=0;B<4;B++){m[B][0]=m[B+4][3]=B&1?x+E:E;m[B][1]=m[B+4][4]=(B>1?v+D:D);m[B][6]=(B&1?-E-x:-E)*(G[B].x+E);m[B][7]=(B>1?-D-v:-D)*(G[B].x+E);m[B+4][6]=(B&1?-E-x:-E)*(G[B].y+D);m[B+4][7]=(B>1?-D-v:-D)*(G[B].y+D);i[B]=(G[B].x+E);i[B+4]=(G[B].y+D);m[B][2]=m[B+4][5]=1;m[B][3]=m[B][4]=m[B][5]=m[B+4][0]=m[B+4][1]=m[B+4][2]=0}var y,r;var u;var s=[];var B,A,z,F;for(var A=0;A<8;A++){for(var B=0;B<8;B++){s[B]=m[B][A]}for(B=0;B<8;B++){u=m[B];y=B<A?B:A;r=0;for(var z=0;z<y;z++){r+=u[z]*s[z]}u[A]=s[B]-=r}var w=A;for(B=A+1;B<8;B++){if(Math.abs(s[B])>Math.abs(s[w])){w=B}}if(w!=A){for(z=0;z<8;z++){F=m[w][z];m[w][z]=m[A][z];m[A][z]=F}F=q[w];q[w]=q[A];q[A]=F}if(m[A][A]!=0){for(B=A+1;B<8;B++){m[B][A]/=m[A][A]}}}for(B=0;B<8;B++){q[B]=i[q[B]]}for(z=0;z<8;z++){for(B=z+1;B<8;B++){q[B]-=q[z]*m[B][z]}}for(z=7;z>-1;z--){q[z]/=m[z][z];for(B=0;B<z;B++){q[B]-=q[z]*m[B][z]}}var C="matrix3d("+q[0].toFixed(9)+","+q[3].toFixed(9)+", 0,"+q[6].toFixed(9)+","+q[1].toFixed(9)+","+q[4].toFixed(9)+", 0,"+q[7].toFixed(9)+",0, 0, 1, 0,"+q[2].toFixed(9)+","+q[5].toFixed(9)+", 0, 1)";if(b.useDPRFix){var o=b.dpr;C="scale("+o+","+o+")perspective(1000px)"+C+"translateZ("+((1-o)*1000)+"px)"}return this.style[f]=C}n();d.update=e;d.checkError=g;return d})();return b})}(typeof define==="function"&&define.amd?define:function(a){window.PerspectiveTransform=a()}));


gmic.php
<?php

$files = glob('tmp/*');
$action = 'list';

if( isset($_POST['action']) && $_POST['action'] != ''){
  $action = $_POST['action'];
}

if($action == 'list'){
  if(count($files) == 0){
    echo 0;
  }  else {
    echo str_replace("tmp/", "", explode(".", $files[0])[0]);
  }
}

// Save image from URL
function getimg($url) {         
    $headers[] = 'Accept: image/gif, image/x-bitmap, image/jpeg, image/pjpeg, image/png';              
    $headers[] = 'Connection: Keep-Alive';         
    $headers[] = 'Content-type: application/x-www-form-urlencoded;charset=UTF-8';         
    $user_agent = 'php';         
    $process = curl_init($url);         
    curl_setopt($process, CURLOPT_HTTPHEADER, $headers);         
    curl_setopt($process, CURLOPT_HEADER, 0);         
    curl_setopt($process, CURLOPT_USERAGENT, $user_agent);
    curl_setopt($process, CURLOPT_TIMEOUT, 30);         
    curl_setopt($process, CURLOPT_RETURNTRANSFER, 1);         
    curl_setopt($process, CURLOPT_FOLLOWLOCATION, 1);         
    $return = curl_exec($process);         
    curl_close($process);         
    return $return;     
} 

// Generate images
if ($action == 'generate') {

  // Clear old images
  foreach($files as $file){
    if(is_file($file)) {
      unlink($file);
    }
  }

  $img_id = date("YmdHis");
  $imgpath = "tmp/${img_id}.jpg";
  if($_POST['isUrl'] == 'true'){
      $imgurl = $_POST['data'];
      copy($imgurl, $imgpath);
  } else {
      $extension = explode('/', mime_content_type($_POST['data']))[1];
      $data = base64_decode(preg_replace('#^data:image/\w+;base64,#i', '', $_POST['data']));
      file_put_contents("tmp/${img_id}.jpg", $data);
  }

  // Posterize
  exec('gmic -input ' . $imgpath . ' -fx_posterize 150,30,1,2,0,0,1,0 -output tmp/' . $img_id . '_pz1.jpg');
  exec('gmic -input ' . $imgpath . ' -fx_posterize 150,30,1,4,0,0,1,0 -output tmp/' . $img_id . '_pz2.jpg');
  exec('gmic -input ' . $imgpath . ' -fx_posterize 150,30,1,6,0,0,1,0 -output tmp/' . $img_id . '_pz3.jpg');
  exec('gmic -input ' . $imgpath . ' -fx_posterize 150,30,1,10,0,0,1,0 -output tmp/' . $img_id . '_pz4.jpg');
  exec('gmic -input ' . $imgpath . ' -fx_posterize 150,30,1,12,0,0,1,0 -output tmp/' . $img_id . '_pz5.jpg');
  exec('gmic -input ' . $imgpath . ' -fx_posterize 150,30,1,16,0,0,1,0 -output tmp/' . $img_id . '_pz6.jpg');
  exec('gmic -input ' . $imgpath . ' -fx_posterize 150,30,1,20,0,0,1,0 -output tmp/' . $img_id . '_pz7.jpg');

  // Vectorize
  exec('gmic -input ' . $imgpath . ' -fx_vector_painting 3 -output tmp/' . $img_id . '_vc1.jpg');
  exec('gmic -input ' . $imgpath . ' -fx_vector_painting 6 -output tmp/' . $img_id . '_vc2.jpg');
  exec('gmic -input ' . $imgpath . ' -fx_vector_painting 7 -output tmp/' . $img_id . '_vc3.jpg');
  exec('gmic -input ' . $imgpath . ' -fx_vector_painting 8 -output tmp/' . $img_id . '_vc4.jpg');
  exec('gmic -input ' . $imgpath . ' -fx_vector_painting 9 -output tmp/' . $img_id . '_vc5.jpg');
  exec('gmic -input ' . $imgpath . ' -fx_vector_painting 9.2 -output tmp/' . $img_id . '_vc6.jpg');
  exec('gmic -input ' . $imgpath . ' -fx_vector_painting 9.4 -output tmp/' . $img_id . '_vc7.jpg');
  exec('gmic -input ' . $imgpath . ' -fx_vector_painting 9.6 -output tmp/' . $img_id . '_vc8.jpg');
  exec('gmic -input ' . $imgpath . ' -fx_vector_painting 9.8 -output tmp/' . $img_id . '_vc9.jpg');
  exec('gmic -input ' . $imgpath . ' -fx_vector_painting 9.9 -output tmp/' . $img_id . '_vc10.jpg');

  // Restore details
  exec('gmic -input ' . $imgpath . ' -fx_freaky_details 2,10,1,11,0,32,0 -output tmp/' . $img_id . '_fd.jpg');

  // Outlines
  exec('gmic -input ' . $imgpath . ' -edges 24% -normalize 0,255 -output tmp/' . $img_id . '_ol1.jpg');
  exec('gmic -input ' . $imgpath . ' -edges 18% -normalize 0,255 -output tmp/' . $img_id . '_ol2.jpg');
  exec('gmic -input ' . $imgpath . ' -edges 14% -normalize 0,255 -output tmp/' . $img_id . '_ol3.jpg');
  exec('gmic -input ' . $imgpath . ' -edges 9% -normalize 0,255 -output tmp/' . $img_id . '_ol4.jpg');
  exec('gmic -input ' . $imgpath . ' -edges 6% -normalize 0,255 -output tmp/' . $img_id . '_ol5.jpg');
  exec('gmic -input ' . $imgpath . ' -edges 4% -normalize 0,255 -output tmp/' . $img_id . '_ol6.jpg');
  exec('gmic -input ' . $imgpath . ' -edges 2% -normalize 0,255 -output tmp/' . $img_id . '_ol7.jpg');
  exec('gmic -input ' . $imgpath . ' -edges 1% -normalize 0,255 -output tmp/' . $img_id . '_ol8.jpg');

  echo $img_id;
}

?>
